### DISCUSSION TITLE
[Application] Provider.Quest

### DISCUSSION THREAD

**Discussion by jimpick** [filecoin-project/community#208] [Project Showcase]
# Provider.Quest (formerly Miner.Report)

Provider.Quest is a project to collect analytics data from the Filecoin blockchain. It is also a collection of notebooks on ObservableHQ with interactive visualizations that use the data.

Access it at https://provider.quest/ (redirects to the Provider.Quest ObservableHQ Collection)

The project is live on [ObservableHQ](https://observablehq.com/collection/@jimpick/provider-quest), with continuously updating deal data (daily and hourly), miner power and storage asks. The core data pipeline, powered by [Apache Spark Structured Streaming](https://spark.apache.org/docs/latest/structured-streaming-programming-guide.html), [Textile Buckets](https://docs.textile.io/buckets/) and [Estuary](https://estuary.tech/), is running 24/7 publishing new data multiple times per hour. The project is in an ideal state for new data sources, reports and visualizations to be added.

<img width="965" alt="miner-report-observable" src="https://user-images.githubusercontent.com/20886/124697210-45f65200-de9b-11eb-8349-97be1f236b38.png">
<img width="617" alt="miner-report-deals-plot" src="https://user-images.githubusercontent.com/20886/124697230-4a226f80-de9b-11eb-9026-8fac37b42ae4.png">

**Comment by jimpick** [comment]
I submitted a microgrant application here to add some geo data to the project: https://github.com/filecoin-project/devgrants/issues/266

I think it might be the first microgrant application ever!

I'll do this eventually, but if I can get funded, I can do it right away. :-)

**Comment by jimpick** [comment]
I'm now publishing data for "named clients" ... I can map client addresses to names and the system will calculate and publish data for those. Here's a new notebook to view the data (showing only the "Estuary" deals). You can toggle "All deals" on and off to see how the named client data compares.

https://observablehq.com/@jimpick/miner-report-deals-named-clients-hourly?name=estuary

<img width="660" alt="Screen Shot 2021-07-08 at 1 04 43 PM" src="https://user-images.githubusercontent.com/20886/124983875-61717200-dfed-11eb-88ae-15a26036ef6f.png">

**Comment by jimpick** [comment]
Great news! I applied for a Filecoin "Next Step" devgrant and it was approved!

"Accurate geolocation data for visualizations"

https://github.com/filecoin-project/devgrants/issues/266

I spent the last few days working on a good system for collecting up-to-date multiaddrs from the on-chain "miner info" data as well as continuously scanning the DHT. As a bonus, I'm doing DNS resolutions on the multiaddrs ... currently only for the on-chain data (nobody seems to be publishing dns-based multiaddrs to the DHT presently).

The next step is to publish the latest DHT data to a Textile bucket - I should have that done today. Then I should have a complete set of IP addresses that is continually updated to do IP geo-location with. There are a few other things I'm going to do as well as part of the grant.

**Reply by jimpick** [reply]
I'm now doing DNS lookups for DHT-based multiaddrs now too. They are being used, but my scripts weren't collecting many of them.

**Reply by jennijuju** [reply]
Nice!

**Reply by jimpick** [reply]
Today I added some aggregations to count the number of times I've had successful DHT queries over a single day and multiday window ... this is just an optimization so I don't spend so much time waiting for entries that probably don't exist. Now the scanning script uses that list and it can do a full scan of the recently seen miners in the DHT in about 3 minutes. I also randomly scan miners that haven't been seen recently to discover any that have come back online. You can see it in action here:

https://observablehq.com/@jimpick/miner-report-dht-miner-peer-scanner?collection=@jimpick/miner-report

**Comment by jimpick** [comment]
Just added to Miner.Report … a new notebook where I take the latest on-chain libp2p addresses + the results from the latest DHT scan, and I extract all the IP addresses (IPv4 and IPv6) … including DNS resolutions for /dns4/ and /dns6/ multi-addresses => https://observablehq.com/@jimpick/miner-report-multiaddr-ip-tool?collection=@jimpick/miner-report

**Comment by jimpick** [comment]
Here’s a quick https://miner.report update … I’ve done a bunch of behind-the-scenes work to optimize the various scanning processes (miner power, miner info, DHT, asks, etc.) to focus on the most-likely-to-be-available data first, followed by randomized “wider focus” scans to capture data from miners outside that initial focus. I decided to ignore miners that have zero power averages … so new miners will need to seal at least a sector before the system will pick them up. Yesterday, I built a scanner notebook that does GeoIP lookups for all of the IP addresses using the MaxMind GeoLite2 database. Today I will put that into Apache Spark and see if I can republish the data to a Textile Bucket. If you are interested, you can play with the scanner notebook here: https://observablehq.com/@jimpick/miner-report-maxmind-geolite2-lookups?collection=@jimpick/miner-report

**Comment by jimpick** [comment]
Now that I’ve got location data in https://miner.report/ - I made a little notebook to explore where the miners are and organize them into geographic “Continent”-“Country” buckets. I also wanted to see how many miners are in multiple regions. I also added the ability to filter against ask/deal information. Nothing visual yet, but it’s fun to play with the data. I’m planning to output some of this so I can prepare aggregated stats by region. https://observablehq.com/@jimpick/miner-report-miner-to-region-mapper

**Comment by jimpick** [comment]
Miner.Report update … I’m now computing aggregated statistics over time for the various “regions” I have bucketed miners into. Only power right now, but I’ll add some deal stats soon. => https://observablehq.com/@jimpick/miner-report-regional-stats?collection=@jimpick/miner-report

**Comment by jimpick** [comment]
I am now breaking out the China-based miner stats by province - I wanted to do that since each province is bigger than many countries. Check it out here => https://observablehq.com/@jimpick/miner-report-regional-stats?region=AS-CN-XJ

**Comment by jimpick** [comment]
Newest enhancement - Regional Deal Stats!! You can now see how many deals are happening in each region, how much data, how many providers, how many clients, etc. => https://observablehq.com/@jimpick/miner-report-regional-stats?region=NA-US

**Comment by Joss-Hua** [comment]
Pretty good. If it can be made into a website and updated every day, it will be a good product for users.

**Comment by jimpick** [comment]
Newest addition to miner.report … a Zoomable “Sunburst” visualization! (Thanks to D3.js) See the network power, provider count, and deal data grouped by region over time. With animation support too! https://observablehq.com/@jimpick/miner-report-zoomable-sunburst?collection=@jimpick/miner-report

![sunburst](https://user-images.githubusercontent.com/20886/130203870-d6441451-3686-4b3d-9031-6cabb9c412c6.gif)

**Comment by jimpick** [comment]
I added a “Weekly Deals” notebook to https://miner.report … it shows some pretty nice growth in deals recently! https://observablehq.com/@jimpick/miner-report-published-deals-weekly?collection=@jimpick/miner-report

<img width="657" alt="Screen Shot 2021-08-25 at 9 33 11 AM" src="https://user-images.githubusercontent.com/20886/130830240-c6a34336-198a-4723-8d3d-ce7362779bfc.png">

**Comment by jimpick** [comment]
The big renaming is done. The new name of the project is Provider.Quest and it can be found at https://provider.quest/ ... same data, just a new name!

**Comment by jimpick** [comment]
I applied for another microgrant to "Add Historical Published Deal Data + Active/Slashed Deals" ... let's see if it's possible to get another one.

https://github.com/filecoin-project/devgrants/issues/397

I'd greatly appreciate any support!

I didn't apply for a larger "open grant" as I'd prefer to just add small bits to this project over time when it makes sense and fits into my schedule.

**Comment by jimpick** [comment]
Provider.Quest now has stats by individual country (plus US states and China provinces).

https://observablehq.com/@jimpick/country-state-province-stats-provider-quest
